diff --git a/dist/development/index-react-server.js b/dist/development/index-react-server.js
index a82f837c8c31138200a1b2d8866b2c4dd88e44e9..1a281fe01ff626dd717eb44636ab65171e854f96 100644
--- a/dist/development/index-react-server.js
+++ b/dist/development/index-react-server.js
@@ -3029,6 +3029,16 @@ async function generateRenderResponse(request, routes, basename, isDataRequest,
         if (isMutationMethod(request.method)) {
           try {
             throwIfPotentialCSRFAttack(request, allowedActionOrigins);
+          } catch (error) {
+            onError?.(error);
+            potentialCSRFAttackError = error;
+            request = new Request(request.url, {
+              method: "GET",
+              headers: request.headers,
+              signal: request.signal
+            });
+          }
+          if (!potentialCSRFAttackError) {
             ctx.runningAction = true;
             let result2 = await processServerAction(
               request,
@@ -3068,13 +3078,11 @@ async function generateRenderResponse(request, routes, basename, isDataRequest,
                 void 0
               );
             }
-          } catch (error) {
-            potentialCSRFAttackError = error;
           }
         }
         let staticContext = await query(
           request,
-          skipRevalidation || !!potentialCSRFAttackError ? {
+          skipRevalidation ? {
             filterMatchesToLoad: () => false
           } : void 0
         );
diff --git a/dist/development/index-react-server.mjs b/dist/development/index-react-server.mjs
index aa5b14390911f1b50ff5084346611cf426361900..3494ff519348f78dafe1d06958e461d8cb233282 100644
--- a/dist/development/index-react-server.mjs
+++ b/dist/development/index-react-server.mjs
@@ -3008,6 +3008,16 @@ async function generateRenderResponse(request, routes, basename, isDataRequest,
         if (isMutationMethod(request.method)) {
           try {
             throwIfPotentialCSRFAttack(request, allowedActionOrigins);
+          } catch (error) {
+            onError?.(error);
+            potentialCSRFAttackError = error;
+            request = new Request(request.url, {
+              method: "GET",
+              headers: request.headers,
+              signal: request.signal
+            });
+          }
+          if (!potentialCSRFAttackError) {
             ctx.runningAction = true;
             let result2 = await processServerAction(
               request,
@@ -3047,13 +3057,11 @@ async function generateRenderResponse(request, routes, basename, isDataRequest,
                 void 0
               );
             }
-          } catch (error) {
-            potentialCSRFAttackError = error;
           }
         }
         let staticContext = await query(
           request,
-          skipRevalidation || !!potentialCSRFAttackError ? {
+          skipRevalidation ? {
             filterMatchesToLoad: () => false
           } : void 0
         );
diff --git a/dist/production/index-react-server.js b/dist/production/index-react-server.js
index a82f837c8c31138200a1b2d8866b2c4dd88e44e9..1a281fe01ff626dd717eb44636ab65171e854f96 100644
--- a/dist/production/index-react-server.js
+++ b/dist/production/index-react-server.js
@@ -3029,6 +3029,16 @@ async function generateRenderResponse(request, routes, basename, isDataRequest,
         if (isMutationMethod(request.method)) {
           try {
             throwIfPotentialCSRFAttack(request, allowedActionOrigins);
+          } catch (error) {
+            onError?.(error);
+            potentialCSRFAttackError = error;
+            request = new Request(request.url, {
+              method: "GET",
+              headers: request.headers,
+              signal: request.signal
+            });
+          }
+          if (!potentialCSRFAttackError) {
             ctx.runningAction = true;
             let result2 = await processServerAction(
               request,
@@ -3068,13 +3078,11 @@ async function generateRenderResponse(request, routes, basename, isDataRequest,
                 void 0
               );
             }
-          } catch (error) {
-            potentialCSRFAttackError = error;
           }
         }
         let staticContext = await query(
           request,
-          skipRevalidation || !!potentialCSRFAttackError ? {
+          skipRevalidation ? {
             filterMatchesToLoad: () => false
           } : void 0
         );
diff --git a/dist/production/index-react-server.mjs b/dist/production/index-react-server.mjs
index aa5b14390911f1b50ff5084346611cf426361900..3494ff519348f78dafe1d06958e461d8cb233282 100644
--- a/dist/production/index-react-server.mjs
+++ b/dist/production/index-react-server.mjs
@@ -3008,6 +3008,16 @@ async function generateRenderResponse(request, routes, basename, isDataRequest,
         if (isMutationMethod(request.method)) {
           try {
             throwIfPotentialCSRFAttack(request, allowedActionOrigins);
+          } catch (error) {
+            onError?.(error);
+            potentialCSRFAttackError = error;
+            request = new Request(request.url, {
+              method: "GET",
+              headers: request.headers,
+              signal: request.signal
+            });
+          }
+          if (!potentialCSRFAttackError) {
             ctx.runningAction = true;
             let result2 = await processServerAction(
               request,
@@ -3047,13 +3057,11 @@ async function generateRenderResponse(request, routes, basename, isDataRequest,
                 void 0
               );
             }
-          } catch (error) {
-            potentialCSRFAttackError = error;
           }
         }
         let staticContext = await query(
           request,
-          skipRevalidation || !!potentialCSRFAttackError ? {
+          skipRevalidation ? {
             filterMatchesToLoad: () => false
           } : void 0
         );
